peer/upstream/upstream.go

```diff
@@ -0,0 +1,350 @@
+		u.log.Errorf("getHeight: %s send error: %s", u.client, err)
+		return 0, err
+	}
+
+	data, err := u.client.Receive(0)
+	if nil != err {
+		u.log.Errorf("push: %s receive error: %s", u.client, err)
+		return 0, err
+	}
+	if 2 != len(data) {
+		return 0, fmt.Errorf("getHeight received: %d  expected: 2", len(data))
+	}
+
+	switch string(data[0]) {
+	case "E":
+		return 0, fmt.Errorf("rpc error response: %q", data[1])
+	case "N":
+		if 8 != len(data[1]) {
+			return 0, fmt.Errorf("highestBlock: rpc invalid response: %q", data[1])
+		}
+		height := binary.BigEndian.Uint64(data[1])
+		u.updateLastResponseTime()
+		u.log.Infof("height: %d", height)
+		return height, nil
+	default:
+		return 0, fmt.Errorf("rpc unexpected response: %q", data[0])
+	}
+}
+
+func (u *Upstream) updateLastResponseTime() {
+	u.lastResponseTime = time.Now()
+}
+
+func (u *Upstream) saveRemoteHeight(height uint64) {
+	u.blockHeight = height
+}
+
+func (u *Upstream) saveDigest(digest blockdigest.Digest) {
+	u.digest = digest
+}
+
+func (u *Upstream) GetSavedDigest() blockdigest.Digest {
+	return u.digest
+}
+
+// start monitoring the disconnect signal on client socket
+func (u *Upstream) monitorDisconnectSig() error {
+	addr := u.client.String()
+	if "" == addr {
+		return fault.InvalidError("invalid address")
+	}
+	sig := fmt.Sprintf(monitorSignal, strings.Replace(addr, "tcp://", "", 1))
+	u.log.Debugf("monitor socket with signal: %q", sig)
+	return u.client.StartMonitoring(sig, zmq.EVENT_DISCONNECTED)
+}
+
+func (u *Upstream) stopPolling() {
+	if nil == u.stopPollingSig {
+		return
+	}
+	u.log.Debug("stopping polling…")
+	close(u.stopPollingSig)
+	u.stopPollingSig = nil
+}
+
+// process the socket events
+func (u *Upstream) handleEvent(socket *zmq.Socket) {
+	ev, addr, v, err := socket.RecvEvent(0)
+	if nil != err {
+		u.log.Errorf("receive event error: %s", err)
+		return
+	}
+	u.log.Debugf("event: %q  address: %q  value: %d", ev, addr, v)
+
+	switch ev {
+	case zmq.EVENT_DISCONNECTED:
+		// reconnect to server
+		u.Lock()
+		u.reconnect()
+		u.Unlock()
+
+	default:
+	}
+}
+
+// reconnect to server
+//
+// need to hold the lock before calling
+func (u *Upstream) reconnect() error {
+
+	u.connected = false
+
+	// stop polling
+	u.stopPolling()
+
+	// try to reconnect
+	u.log.Infof("reconnecting to [%s]…", u.client.String())
+	err := u.client.Reconnect()
+	if nil != err {
+		u.log.Errorf("reconnect to [%s] error: %s", u.client.String(), err)
+		return err
+	}
+
+	u.log.Infof("reconnect to [%s] successfully", u.client.String())
+
+	// start polling again after reconnect
+	stopPollingSig := make(chan struct{})
+	go u.startPolling(stopPollingSig)
+	u.stopPollingSig = stopPollingSig
+
+	return nil
+}
+
+func (u *Upstream) RemoteAddr() (string, error) {
+	var err error
+
+	if nil == u.client {
+		err = fmt.Errorf("client not created")
+	} else if !u.client.IsConnected() {
+		err = fmt.Errorf("client socket not connected")
+	}
+
+	if nil != err {
+		u.log.Errorf(err.Error())
```

Comment:
`errorf` with single params is still valid?

Reply:
can you test that?

Reply:
No. Just question you. It still work?

Reply:
@jamieabc "can you test that?" Don't you test your code before commit and open a pr and ask for review? You must test your own code and not ask reviewer to do the test for you.

Reply:
I think if it's a suggestion, commenter one should have clear confirm about comment that something should world.

And yes, I have test, it works

Reply:
Why don't you use `u.log.Error()` instead of `u.log.Errorf()` in case you only pass 1 params?

Reply:
yes, it can be changed